Implement Loadbalancer, Listener, and TargetGroup from elasticloadbalancingv2

EdgeJ opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Implements 3 resource types from the `elasticloadbalancingv2` api group:
* LoadBalancer
* Listener
* TargetGroup


Fixes #841 
Fixes #699

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->
There are 3 example yaml files to generate the new resources.  The listeners.yaml file contains specifications to create 2 listeners and attach them to the same loadbalancer.  Additionally, there are unit tests for the function used to generate `DefaultActions` types from the API that will work with the AWS SDK.  This function was largely adapted from autogenerated code, but was needed in order to implement references within the API type.

[contribution process]: https://git.io/fj2m9

EdgeJ commented on pkg/controller/elasticloadbalancingv2/loadbalancer/setup.go:
```diff
@@ -0,0 +1,81 @@
+		obj.Names = []*string{aws.String(meta.GetExternalName(cr))}
+	} else {
+		obj.LoadBalancerArns = []*string{aws.String(meta.GetExternalName(cr))}
+	}
```
This is great! I hadn't realized about the `WithInitializers` option before.  This helps solve few problems I had: namely, the fact that API calls failed when passed strings that weren't ARNs and the references to other objects returning non-ARN strings and breaking the ability to apply LoadBalancers and Listeners simultaneously. I obviously hacked my way around the former here, but after testing adding `WithInitializers` I have removed this.

haarchri commented:
@EdgeJ The code-generator now supports setting a different "model name" as part of the generator.yaml file : https://github.com/aws-controllers-k8s/community/issues/985 - we need to bump to latest Code-generator

EdgeJ commented:
@haarchri following the advice in that thread doesn't seem to be working for me.  I updated the tag used for the code generator to `v0.15.2` (current latest) and added `module_name: elasticloadbalancingv2` to `generator-config.yaml`.  When attempting to generate the service, I get nil pointer errors:
```
➜ SERVICES=elasticloadbalancingv2 make services
HEAD is now at d2b0638 Update to ACK runtime `v0.15.2` (#234)
11:58:12 [ .. ] Generating elasticloadbalancingv2 controllers and CRDs
Error: template: /Users/jedge/git/crossplane/provider-aws/.work/code-generator/templates/crossplane/pkg/conversions.go.tpl:40:3: executing "/Users/jedge/git/crossplane/provider-aws/.work/code-generator/templates/crossplane/pkg/conversions.go.tpl" at <GoCodeSetUpdateInput .CRD "cr" "res" 1>: error calling GoCodeSetUpdateInput: runtime error: invalid memory address or nil pointer dereference
exit status 1
make: *** [services] Error 1
```

I'll keep debugging for now, but if you have any insight, it would help greatly.

EdgeJ commented:
I discovered the issue.  For the `Listener` objects in particular, generation is broken by ACK metadata.  Adding a configuration line `include_ack_metadata: false` seems to fix the nil pointer issue.  A bigger problem is that the `crossplane` command for the code-generator does not honor the `model_name` configuration parameter.  I added a PR to the repo to hopefully amend this: https://github.com/aws-controllers-k8s/code-generator/pull/237

janwillies commented:
this is blocked by https://github.com/crossplane/provider-aws/pull/920

haarchri commented:
@EdgeJ  we resolved #920 - did you have time to rebase and check again ?

EdgeJ commented:
> @EdgeJ we resolved #920 - did you have time to rebase and check again ?

Thanks @haarchri.  Will take a look and build this with the latest code-generator.

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Reference to Security Groups for SecurityGroups field
+	// +optional
+	SecurityGroupsRef []xpv1.Reference `json:"securityGroupsRef,omitempty"`
```
`SecurityGroupRefs`

We usually go for plural in Refs instead of the field in case of arrays.

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Reference to Subnets for Subnets field
+	// +optional
+	SubnetsRef []xpv1.Reference `json:"subnetsRef,omitempty"`
```
`SubnetRefs`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+type CustomTargetGroupParameters struct {
+	// Reference to VPC for VPCID
+	// +optional
+	VPCIDRef *xpv1.Reference `json:"vpcIDRef,omitempty"`
```
`json:"vpcIdRef,omitempty`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Selector for references to VPC for VPCID
+	// +optional
+	VPCIDSelector *xpv1.Selector `json:"vpcIDSelector,omitempty"`
```
`json:"vpcIdSelector,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// The Amazon Resource Name (ARN) of the load balancer.
+	// +optional
+	LoadBalancerARN *string `json:"loadBalancerARN,omitempty"`
```
`json:"loadBalancerArn,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Ref to loadbalancer ARN
+	// +optional
+	LoadBalancerARNRef *xpv1.Reference `json:"loadBalancerARNRef,omitempty"`
```
`json:"loadBalancerArnRef,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Selector for references to LoadBalancer for LoadBalancerARN
+	// +optional
+	LoadBalancerARNSelector *xpv1.Selector `json:"loadBalancerARNSelector,omitempty"`
```
`json:"loadBalancerArnSelector,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+	// actionType is forward and you want to route to a single target group.
+	// To route to one or more target groups, use ForwardConfig instead.
+	// +optional
+	TargetGroupARN *string `json:"targetGroupARN,omitempty"`
```
`json:"targetGroupArn,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Reference to TargetGroupARN used to set TargetGroupARN
+	// +optional
+	TargetGroupARNRef *xpv1.Reference `json:"targetGroupARNRef,omitempty"`
```
`json:"targetGroupArnRef,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Selector for references to TargetGroups for TargetGroupARNs
+	// +optional
+	TargetGroupARNSelector *xpv1.Selector `json:"targetGroupARNSelector,omitempty"`
```
`json:"targetGroupArnSelector,omitempty"`

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Reference to TargetGroupARN used to set TargetGroupARN
+	// +optional
+	TargetGroupARNRef *xpv1.Reference `json:"targetGroupARNRef,omitempty"`
```
`json:"targetGroupArnRef,omitempty"`

Nitpick: while we don't have explicit convention around this, I think it's fair to go with calculated lower camel, i.e.` targetGroupArnRef`, instead of taking known words into account. There are some places we do one and other places we do the other, but I believe calculated lower camel is the most consistent for repos where there is code generation, like this one or Jet-based providers.

EdgeJ replied:
Yeah, I mostly followed the AWS Go api naming scheme, but I personally prefer lower camel for the CRDs as I find it more readable.

haarchri commented on apis/elbv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,142 @@
+
+	// Selector for references to TargetGroup for TargetGroupARN
+	// +optional
+	TargetGroupARNSelector *xpv1.Selector `json:"targetGroupARNSelector,omitempty"`
```
`json:"targetGroupArnSelector,omitempty"`

haarchri reviewed (changes requested):
thanks for contribution ;) i tested the resource examples - all are working creation/deletion- check my review - after that squash commits for clean history and we can merge the PR

haarchri commented on Makefile:
```diff
@@ -1,3 +1,4 @@
+
```
remove the newline

EdgeJ commented:
@haarchri I updated CRD field names and cleaned up the mistaken newline in `Makefile`.  One caveat is that `make reviewable` seems to be failing on code generation for the `sfn` group, but this seems to be a regression of some sort in `master` and not related to these changes.  Removing `sfn` from the `GENERATED_SERVICES` variable allows `make reviewable` to pass.

haarchri commented:
i will start review

haarchri reviewed (approved):
This works at creating and deleting:

```
NAME                                                   READY   SYNCED   EXTERNAL-NAME
targetgroup.elbv2.aws.crossplane.io/test-targetgroup   True    True     arn:aws:elasticloadbalancing:us-east-1:255932642927:targetgroup/test-targetgroup/cbae8e0ba0ebb033

NAME                                                 READY   SYNCED   EXTERNAL-NAME
listener.elbv2.aws.crossplane.io/test-listener       True    True     arn:aws:elasticloadbalancing:us-east-1:255932642927:listener/app/test-loadbalancer/d1441d261a57f64d/bce43527425177dc
listener.elbv2.aws.crossplane.io/test-listener-two   True    True     arn:aws:elasticloadbalancing:us-east-1:255932642927:listener/app/test-loadbalancer/d1441d261a57f64d/9853d163a3de165f

NAME                                                     READY   SYNCED   EXTERNAL-NAME
loadbalancer.elbv2.aws.crossplane.io/test-loadbalancer   True    True     arn:aws:elasticloadbalancing:us-east-1:255932642927:loadbalancer/app/test-loadbalancer/d1441d261a57f64d
```

delete load balancer before listener results in 
`failed to describe Listener: LoadBalancerNotFound: Load balancer 'arn:aws:elasticloadbalancing:us-east-1:255932642927:loadbalancer/app/test-loadbalancer/972b04e3dc92fd7d' not found`-
 I do not think this needs to be solved for alpha quality - but we need to check it via followup PR
